eis_toolkit/prediction/weights_of_evidence.py

```diff
@@ -0,0 +1,255 @@
+    A = np.sum(np.logical_and(deposits == 1, evidence == 1))
+    B = np.sum(np.logical_and(deposits == 1, evidence == 0))
+    C = np.sum(np.logical_and(deposits == 0, evidence == 1))
+    D = np.sum(np.logical_and(deposits == 0, evidence == 0))
+
+    if A + B == 0:
+        raise Exception("No deposits")
+    if C + D == 0:
+        raise Exception("All included cells have deposits")
+
+    if A == 0:
+        return A, B, C, D, 0, 0, 0, 0, 0, 0, 0
+
+    p_A_nominator = A
+    p_C_nominator = C
+    B_adjusted = B
+    D_adjusted = D
+
+    if B == 0:
+        p_A_nominator -= 0.99
+        B_adjusted = 0.99
+
+    if D == 0:
+        p_C_nominator -= 0.99
+        D_adjusted = 0.99
+
+    p_A = p_A_nominator / (A + B)  # probability of presence of evidence given the presence of mineral deposit
+    p_C = p_C_nominator / (C + D)  # probability of presence of evidence given the absence of mineral deposit
+
+    # Calculate metrics
+    w_plus = np.log(p_A / p_C) if p_C != 0 else 0  # Check
+    w_minus = np.log((1 - p_A) / (1 - p_C))
+    contrast = w_plus - w_minus
+
+    # Calculate signifigance metrics
+    s_w_plus = np.sqrt((1 / p_A_nominator) + (1 / p_C_nominator))
+    s_w_minus = np.sqrt((1 / B_adjusted) + (1 / D_adjusted))
+
+    s_contrast = np.sqrt(s_w_plus**2 + s_w_minus**2)
+    studentized_contrast = contrast / s_contrast
+
+    return A, B, C, D, w_plus, s_w_plus, w_minus, s_w_minus, contrast, s_contrast, studentized_contrast
+
+
+def _unique_weights(deposits: np.ndarray, evidence: np.ndarray) -> dict:
+    """Calculate unique weights for each class."""
+    classes = np.unique(evidence)
+    return {cls: _calculate_metrics_for_class(deposits, evidence == cls) for cls in classes}
+
+
+def _cumulative_weights(deposits: np.ndarray, evidence: np.ndarray, ascending: bool = True) -> dict:
+    """Calculate cumulative weights (ascending or descending) for each class."""
+    classes = sorted(np.unique(evidence), reverse=not ascending)
+    cumulative_classes = [classes[: i + 1] for i in range(len(classes))]
+    return {
+        cls[i]: _calculate_metrics_for_class(deposits, np.isin(evidence, cls))
+        for i, cls in enumerate(cumulative_classes)
+    }
+
+
+def _reclassify_by_studentized_contrast(df: pd.DataFrame, studentized_contrast_threshold: Number) -> None:
+    """Create generalized classes based on the studentized contrast threhsold value."""
+    index = df.idxmax()["Contrast"]
+
+    if df.loc[index, "Studentized contrast"] < studentized_contrast_threshold or index == len(df.index) - 1:
+        raise exceptions.ClassificationFailedException(
+            "Failed to create generalized classes with given studentized contrast treshold ({})".format(
+                df.loc[index, "Studentized contrast"]
+            )
+        )
+
+    df["Generalized class"] = 1
+    for i in range(0, index + 1):
+        df.loc[i, "Generalized class"] = 2
+
+
+def _calculate_generalized_weights(df: pd.DataFrame, deposits) -> None:
```

Comment:
```suggestion
def _calculate_generalized_weights(df: pd.DataFrame, deposits) -> pd.DataFrame:
```

I would suggest making a copy of the input dataframe at the start of the function or at the very least returning the modified dataframe at the end for readability. Functions with side effects such as this should be avoided in my opinion but that is sadly the style of `pandas` rather than our choice so these are unavoidable to some extent.